### PR TITLE
Make MooseRandom restartable

### LINKED ISSUE #8225: Make MooseRandom restartable
### Description of the enhancement or error report
The state of the MooseRandom system cannot currently be carried over from run to run. Restarted/recovered simulations will start over at the beginning of the random number sequence. There is no system for reading/writing the information about the `mt_state` type to file. Doing this *should* allow the system to be restartable.

### Rationale for the enhancement or information for reproducing the error
Better restart/recover behavior.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
None, additional capacity.

See discussion:
https://groups.google.com/forum/?hl=en#!topic/moose-users/pJyJUHc5eac

### PR DESCRIPTION
**Pull request by jessecarterMOOSE**
This appears to work, though I'm a little surprised.

I started by adding dataLoad and dataStore members in DataIO for the MooseRandom class, then just kept adding more store/load members until all the compilation errors went away. The thing is, I never added anything for the mtstate class that mtwist uses. I guess the existing code is smart enough to figure it out on its own? If so, bravo!

I added a test object, RandomPPS, to test this. There's a test that just computes 10 random numbers and then another set of tests that explicitly does a --half-transient and --recover and uses the same gold file. I suppose the latter test isn't required (could just send --recover to the test harness).

For users to use this, it requires them to make their own instance of the MooseRandom class, declare it restartable, and use the non-static calls. Guess that means the static MooseRandom calls are not restartable?

tagging @permcody 

closes #8225

### CODE REVIEW

File: framework/include/utils/MooseRandom.h [hunk 1]
```diff
@@ -167,6 +167,11 @@ class MooseRandom
       it->second.first = it->second.second;
   }
 
+  LIBMESH_BEST_UNORDERED_MAP<unsigned int, std::pair<mt_state, mt_state> > getStates() { return _states; }
```
Comment (dschwen) [inline framework/include/utils/MooseRandom.h]: No. You have not implemented a helper for `LIBMESH_BEST_UNORDERED_MAP`, have you? This only works if this symbol is set to `std::unordered_map`.

**Review by dschwen — CHANGES_REQUESTED** [review]

File: framework/include/restart/DataIO.h [hunk 2]
```diff
@@ -256,6 +269,27 @@ dataStore(std::ostream & stream, std::map<T,U> & m, void * context)
...
+dataStore(std::ostream & stream, std::unordered_map<T,U> & m, void * context)
+{
+  // First store the size of the map
+  unsigned int size = m.size();
+  stream.write((char *) &size, sizeof(size));
+
+  typename std::unordered_map<T,U>::iterator it = m.begin();
+  typename std::unordered_map<T,U>::iterator end = m.end();
+
+  for (; it != end; ++it)
```
Comment (jessecarterMOOSE) [inline framework/include/restart/DataIO.h]: That explicit for loop is used throughout this file for maps. Do you suggest changing all them?

Comment (dschwen) [inline framework/include/restart/DataIO.h]: I'd say go for it.

File: framework/include/utils/MooseRandom.h (hunk 1 above)
Comment (jessecarterMOOSE) [inline framework/include/utils/MooseRandom.h]: Oh, can it be set to different things? Yes, it's `std::unordered_map` on my machine. Is a single helper for `LIBMESH_BEST_UNORDERED_MAP` sufficient or is a helper needed for all possible settings? Or is `LIBMESH_BEST_UNORDERED_MAP` the wrong choice in MooseRandom.h?

Comment (dschwen) [inline framework/include/utils/MooseRandom.h]: Yes, it is set at libmesh configure time (check `libmesh/m4/unordered.m4` for the logic). There are five(!) different possible candidates.

**Comment by jessecarterMOOSE** [comment]
I tried to address @dschwen 's comments but couldn't get the `auto` loop to work. I started getting diffs and a failure in a few of the restart-based tests. I could have tried to narrow it down but every time I make a change to a header file it takes 2+ hours to re-compile on my dinky laptop. Since this PR addresses the _content_ of what I wanted to accomplish, I'm submitting as-is. If `auto` is really the direction that is wanted, then I suggest that be done in a different PR.

**Comment by dschwen** [comment]
Hey Jesse, sorry I'm causing this trouble for you. I can check out your
branch, make the changes and make a new PR for it.

**Comment by friedmud** [comment]
I agree with @jessecarterMOOSE ... let's merge this as is and then do a separate Issue/PR for updating the dataLoad/Store methods to C++11.  That should be done in a separate step anyway to ensure that proper testing is done and it's outside of the purview of what this PR is trying to achieve.

BTW: I'm not giving the green-light to this PR yet... I haven't looked at the code yet... but the idea is good.

File: framework/src/restart/DataIO.C [hunk 3]
```diff
@@ -416,3 +425,12 @@ dataLoad(std::istream & stream, std::stringstream * & s, void * context)
 {
   dataLoad(stream, *s, context);
 }
+
+template<>
+void
+dataLoad(std::istream & stream, MooseRandom & v, void * context)
+{
+  LIBMESH_BEST_UNORDERED_MAP<unsigned int, std::pair<mt_state, mt_state> > states;
+  loadHelper(stream, states, context);
```
Comment (dschwen) [inline framework/src/restart/DataIO.C]: The common pattern is to make the dataLoad and dataStore methods _friends_ of the class. That way you can remove the getter and setter.

**Comment by jessecarterMOOSE** [comment]
It's not a big deal @dschwen. I do enjoy poking around the framework and making changes myself.

**Comment by jessecarterMOOSE** [comment]
Trying this again following what @dschwen suggested about the friend functions.

I poked around the framework to find more examples of how this is done in other classes. I ended up borrowing what's done in `FormattedTable`, only I had to make the functions inline in `MooseRandom.h` since there is no `MooseRandom.C`. It does make things more simpler in that the getter/setter in `MooseRandom` isn't needed, so `DataIO.C` remains untouched and `DataIO.h` only requires an extra store/load function for unordered_maps.

Also since there was some trial and error involved with this PR I don't mind squashing my commits into something more meaningful or redo-ing this PR entirely.

Again I'm a little out of my element here so if this isn't the right way to go about things then let me know. I'm really just naively copying things here...

File: framework/include/restart/DataIO.h [hunk 4]
```diff
@@ -28,6 +28,7 @@
 #ifdef LIBMESH_HAVE_CXX11_TYPE_TRAITS
 #  include <type_traits>
 #endif
+#include LIBMESH_INCLUDE_UNORDERED_MAP
```
Comment (permcody) [inline framework/include/restart/DataIO.h]: We can go ahead and drop all this libmesh best unordered map stuff. It's only here because unordered maps aren't in C++03 and MooseRandom predates our C++11 requirement. Rather than propogate this old code into new parts of MOOSE, I'd rather just switch MooseRandom to std::unordered_set<> directly.

File: framework/include/utils/MooseRandom.h [hunk 5]
```diff
@@ -175,6 +176,13 @@ class MooseRandom
...
    */
   LIBMESH_BEST_UNORDERED_MAP<unsigned int, std::pair<mt_state, mt_state> > _states;
+
+  // for restart capability
+  friend void dataStore<MooseRandom>(std::ostream & stream, MooseRandom & v, void * context);
+  friend void dataLoad<MooseRandom>(std::istream & stream, MooseRandom & v, void * context);
 };
 
+template<> inline void dataStore(std::ostream & stream, MooseRandom & v, void * context) { storeHelper(stream, v._states, context); }
+template<> inline void dataLoad(std::istream & stream, MooseRandom & v, void * context) { loadHelper(stream, v._states, context); }
```
Comment (permcody) [inline framework/include/utils/MooseRandom.h]: This is OK, but I'm also open to adding MooseRandom.C.

File: test/tests/postprocessors/random_pps/tests [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    type = CSVDiff
+    input = random_pps.i
+    csvdiff = random_pps_out.csv
+    cli_args = '--recover'
+    recover = false
+    prereq = random_pps_half_transient
+    delete_output_before_running = false
+    group = random_pps
+  [../]
+[]
```
Comment (permcody) [inline test/tests/postprocessors/random_pps/tests]: Thanks for the thorough testing, however we always run recover tests automatically with moose/test so we don't need the explicit versions of these two part1, part2 tests as you have here. I'm ok with leaving them, just letting you know.

File: framework/include/restart/DataIO.h (hunk 4 above)
Comment (friedmud) [inline framework/include/restart/DataIO.h]: Let's do that as a separate PR.  This PR does what it's intended to do.

File: test/tests/postprocessors/random_pps/tests (hunk 6 above)
Comment (friedmud) [inline test/tests/postprocessors/random_pps/tests]: I'm ok with explicit recovery tests when we're adding explicit recovery code like this.  Most people don't personally run the test suite with `--recover` very often.

**Comment by friedmud** [comment]
I think this is good to go 👍

**Comment by jessecarterMOOSE** [comment]
Naturally I agree with @friedmud simply because it's less work on my part :smile: . Also because I'd like to get this merged sooner so I can focus on what I really need to do, which is make my App restartable.

**Review by dschwen — APPROVED** [review]

**Comment by jessecarterMOOSE** [comment]
Awesome. Thanks everybody. :+1:
